### PR TITLE
[HW5] Add tests to HW2 & HW3

### PR DESCRIPTION
**Pull request by DashaSledopit**

### CODE REVIEW

File: HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_02_spec.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      end
+
+      it 'returns an array with formatted logs' do
+        expect(task2(logs)).to eq formatted_logs
+      end
+
+      it 'ignores error lines' do
+        expect(task2(error_log)).to eq([])
+      end
+    end
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_02_spec.rb:32]:
```suggestion
    context 'when logs are present' do
      context 'when logs are valid' do
         let(:logs) do
           <<~LOGS
            10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0498
            10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] 'POST /test/2/run HTTP/1.1' 200 - 0.2277
            2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>
            10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0290
            LOGS
          end

         let(:formatted_logs) do
           ['23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES',
           '23/Apr/2018:20:30:42 +0300 FROM: 10.6.246.101 TO: /TEST/2/RUN',
           '23/Apr/2018:20:31:39 +0300 FROM: 10.6.246.101 TO: /TEST/2/MESSAGES']
         end
        
         it 'returns an array with formatted logs' do
           expect(task2(logs)).to eq formatted_logs
         end
      end
      
      context 'when logs are not valid' do
        let(:error_log) do
          '2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>'
        end

        it 'ignores error lines' do
          expect(task2(error_log)).to eq([])
        end
      end
    end
```

File: HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_03_spec.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        '2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages'
+      ]
+    end
+
+    let(:invalid_events) do
+      [
+        '2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101',
+        '2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event'
+      ]
+    end
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_03_spec.rb:51]:
```suggestion
```

File: HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_03_spec.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    let(:empty_events) { [] }
+
+    context 'when events are valid' do
+      it 'returns a duration between the events' do
+        expect(calculate_diff(valid_events)).to eq('49.1')
+      end
+    end
+
+    context 'when events of appropriate format are invalid' do
+      it 'returns a duration between the events' do
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_03_spec.rb:62]: add invalid let here please

File: HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_04_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+require './Nastja_Kireeva_hw_03_t_04'
+
+RSpec.describe 'Task4' do
+  describe '.task4' do
+    context 'when string is given' do
+      it 'returns the count of letters and digits in the string' do
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/spec/Nastja_Kireeva_hw_03_t_04_spec.rb:6]:
```suggestion
      it 'returns the count of letters and digits in the string', :aggregate_failures do
```
https://relishapp.com/rspec/rspec-core/docs/expectation-framework-integration/aggregating-failures

File: HW02/Nastja_Kireeva/spec/array_module_spec.rb [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    context 'when block is given' do
+      context 'when array is empty' do
+        it 'returns an empty array' do
+          expect(empty_array.my_map(&:even?)).to be_empty
+        end
+      end
+      context 'when array is not empty' do
+        it 'returns a modified array' do
+          expect(array.my_map { |element| element * 3 }).to eq new_array
+        end
```
Comment (vladhilko) [inline HW02/Nastja_Kireeva/spec/array_module_spec.rb:33]: If you use `new_array` only inside this context make sence to define it only here
```suggestion
      context 'when array is not empty' do
        let(:new_array) { [3, 6, 9, 12, 15, 18] }
        it 'returns a modified array' do
          expect(array.my_map { |element| element * 3 }).to eq new_array
        end
```

**Review by vladhilko — APPROVED** [review]

**Pull request merged**